_posts/2016-12-12-Reactive-Programming-Spring-Reactor.md

```diff
@@ -0,0 +1,525 @@
+* **Resilient**: remain responsive in the event of failure, failures are contained with each component isolating components from each other.
+* **Elastic**: stay responsive under varying workload, reactive systems can react to changes in the input rate by increasing or decreasing the resources allocated to services.
+* **Message Driven**: rely on asynchronous message-passing to establish a boundary between components that ensures loose coupling, isolation and location transparency.
+
+This boundary also provides the means to delegate failures as messages.
+
+Systems built as reactive systems are thus more flexible, loosely-coupled and scalable. 
+This makes them easier to develop and to allow changes.
+They are significantly more tolerant of failure and when failure does occur they meet it with elegance rather than disaster.
+
+
+<a name="latency" />
+
+# Latency
+
+Latency is a real issue, the real physical distance of various components and services becomes more important with cloud based systems.
+This is also a very random number which is difficult to predict because it can depend on network congestion.
+With [Zipkin](https://github.com/openzipkin/zipkin), you can measure this latency.
+
+The same latency can also exist within an application - between the different threads - although the impact will be less severe than between various components.
+
+Something needs to be done when latency becomes too big of an issue, especially if the receiver can not process enough.
+Too much data will fill up the buffer and can result, with an unbounded queue, to the infamous [OutOfMemoryException()](https://docs.oracle.com/javase/8/docs/api/java/lang/OutOfMemoryError.html)
+While you won't run out of memory with a circular buffer, you risk losing messages as the oldest ones get overwritten.
+
+
+# Blocking
+
+One way to prevent out of memory exceptions is to use blocking.
+
+But this can be a real `poison pill`: when a queue is full, it will block a thread and as more and more queues get blocked your server will die a slow death.
+
+Blocking is faster and has better performance, than reactive, but reactive will allow for more concurrency.
+Concurrency is important if you have a microservice based architecture, as there you typically need to be more careful and more exact when allocating resources between services.
+
+As in, by being more concurrent you can save a lot of money when using cloud and microservices.
+
+<a name ="contract" />
+
+# Contract
+
+Reactive is non-blocking and messages will never overflow the queue, see for the standard definition [http://www.reactive-streams.org/](http://www.reactive-streams.org/).
+
+>Created by Pivotal, Typesafe, Netflix, Oracle, Red Hat and others.
+
+The scope of Reactive Streams is to find a minimal set of interfaces, methods and protocols that will describe the necessary operations and entities to achieve the goal—asynchronous streams of data with non-blocking back pressure.
```

Comment:
Maybe also explain back pressure here (or link)? It is referenced later, but never explained.